### ISSUE TITLE
Fails to install on Ubuntu 18.04

### ISSUE THREAD

**Issue by VVelox** [nabijaczleweli/cargo-update#129]
'cargo install cargo-update' results in the item below.  libssh-dev, libssh2-1-dev, or libssh-gcrypt-dev being installed does not solve this.

```
          /tmp/cargo-installxakvhf/release/deps/liblibssh2_sys-c67c8b6af4c7501d.rlib(mac.o): In function `mac_method_hmac_sha2_512_hash':
          (.text+0x43e): undefined reference to `gcry_md_open'
          (.text+0x450): undefined reference to `gcry_md_setkey'
          (.text+0x462): undefined reference to `gcry_md_write'
          (.text+0x471): undefined reference to `gcry_md_write'
          (.text+0x48f): undefined reference to `gcry_md_write'
          (.text+0x499): undefined reference to `gcry_md_get_algo'
          (.text+0x4a0): undefined reference to `gcry_md_get_algo_dlen'
          (.text+0x4ae): undefined reference to `gcry_md_read'
          (.text+0x4c5): undefined reference to `gcry_md_close'
          /tmp/cargo-installxakvhf/release/deps/liblibssh2_sys-c67c8b6af4c7501d.rlib(mac.o): In function `mac_method_hmac_sha2_256_hash':
          (.text+0x53e): undefined reference to `gcry_md_open'
          (.text+0x550): undefined reference to `gcry_md_setkey'
          (.text+0x562): undefined reference to `gcry_md_write'
          (.text+0x571): undefined reference to `gcry_md_write'
          (.text+0x58f): undefined reference to `gcry_md_write'
          (.text+0x599): undefined reference to `gcry_md_get_algo'
          (.text+0x5a0): undefined reference to `gcry_md_get_algo_dlen'
          (.text+0x5ae): undefined reference to `gcry_md_read'
          (.text+0x5c5): undefined reference to `gcry_md_close'
          /tmp/cargo-installxakvhf/release/deps/liblibssh2_sys-c67c8b6af4c7501d.rlib(crypt.o): In function `crypt_dtor':
          (.text+0x1e): undefined reference to `gcry_cipher_close'
          collect2: error: ld returned 1 exit status

```

**Comment by fragrans** [comment]
some here.

**Comment by johanwiden** [comment]
Same issue in Ubuntu 19.10.

**Comment by nabijaczleweli** [comment]
Okay, so: here's my log in a clean chroot for eoan and bionic ([this](https://wiki.ubuntu.com/Releases) Ubuntu site tells me that's "19.10" and "18.04.4 LTS", respectively):
```
root@eoan:/# cargo install cargo-update
    Updating crates.io index
  Installing cargo-update v3.0.0
   Compiling libc v0.2.67
   Compiling pkg-config v0.3.17
   Compiling autocfg v1.0.0
   Compiling maybe-uninit v2.0.0
   Compiling proc-macro2 v1.0.9
   Compiling cfg-if v0.1.10
   Compiling unicode-xid v0.2.0
   Compiling matches v0.1.8
   Compiling memchr v2.3.3
   Compiling bitflags v1.2.1
   Compiling serde v1.0.104
   Compiling syn v1.0.16
   Compiling log v0.4.8
   Compiling unicode-width v0.1.7
   Compiling embed-resource v1.3.2
   Compiling lazy_static v1.4.0
   Compiling percent-encoding v2.1.0
   Compiling regex-syntax v0.6.17
   Compiling strsim v0.8.0
   Compiling vec_map v0.8.1
   Compiling ansi_term v0.11.0
   Compiling semver-parser v0.7.0
   Compiling openssl-probe v0.1.2
   Compiling array_tool v1.0.3
   Compiling lazysort v0.2.1
   Compiling hex v0.4.2
   Compiling json v0.11.15
   Compiling unicode-bidi v0.3.4
   Compiling thread_local v1.0.1
   Compiling textwrap v0.11.0
   Compiling tabwriter v1.2.1
   Compiling cargo-update v3.0.0
   Compiling aho-corasick v0.7.10
   Compiling smallvec v0.6.13
   Compiling unicode-normalization v0.1.9
   Compiling quote v1.0.3
   Compiling jobserver v0.1.21
   Compiling dirs-sys v0.3.4
   Compiling atty v0.2.14
   Compiling clap v2.33.0
   Compiling dirs v2.0.2
   Compiling idna v0.2.0
   Compiling cc v1.0.50
   Compiling url v2.1.1
   Compiling regex v1.3.5
   Compiling libz-sys v1.0.25
   Compiling openssl-sys v0.9.54
   Compiling libssh2-sys v0.2.16
   Compiling libgit2-sys v0.10.0
   Compiling serde_derive v1.0.104
   Compiling toml v0.5.6
   Compiling semver v0.9.0
   Compiling git2 v0.11.0
    Finished release [optimized] target(s) in 52.44s
  Installing /root/.cargo/bin/cargo-install-update
  Installing /root/.cargo/bin/cargo-install-update-config
   Installed package `cargo-update v3.0.0` (executables `cargo-install-update`, `cargo-install-update-config`)
```
```
root@bionic:/# cargo install cargo-update
    Updating crates.io index
  Installing cargo-update v3.0.0
   Compiling libc v0.2.67
   Compiling pkg-config v0.3.17
   Compiling autocfg v1.0.0
   Compiling maybe-uninit v2.0.0
   Compiling proc-macro2 v1.0.9
   Compiling cfg-if v0.1.10
   Compiling memchr v2.3.3
   Compiling serde v1.0.104
   Compiling bitflags v1.2.1
   Compiling unicode-xid v0.2.0
   Compiling matches v0.1.8
   Compiling syn v1.0.16
   Compiling unicode-width v0.1.7
   Compiling log v0.4.8
   Compiling embed-resource v1.3.2
   Compiling lazy_static v1.4.0
   Compiling percent-encoding v2.1.0
   Compiling ansi_term v0.11.0
   Compiling semver-parser v0.7.0
   Compiling regex-syntax v0.6.17
   Compiling vec_map v0.8.1
   Compiling strsim v0.8.0
   Compiling openssl-probe v0.1.2
   Compiling hex v0.4.2
   Compiling json v0.11.15
   Compiling lazysort v0.2.1
   Compiling array_tool v1.0.3
   Compiling unicode-bidi v0.3.4
   Compiling thread_local v1.0.1
   Compiling textwrap v0.11.0
   Compiling tabwriter v1.2.1
   Compiling cargo-update v3.0.0
   Compiling aho-corasick v0.7.10
   Compiling smallvec v0.6.13
   Compiling unicode-normalization v0.1.9
   Compiling quote v1.0.3
   Compiling jobserver v0.1.21
   Compiling atty v0.2.14
   Compiling dirs-sys v0.3.4
   Compiling clap v2.33.0
   Compiling dirs v2.0.2
   Compiling cc v1.0.50
   Compiling idna v0.2.0
   Compiling url v2.1.1
   Compiling regex v1.3.5
   Compiling libz-sys v1.0.25
   Compiling openssl-sys v0.9.54
   Compiling libssh2-sys v0.2.16
   Compiling libgit2-sys v0.10.0
   Compiling serde_derive v1.0.104
   Compiling toml v0.5.6
   Compiling semver v0.9.0
   Compiling git2 v0.11.0
    Finished release [optimized] target(s) in 52.36s
  Installing /root/.cargo/bin/cargo-install-update
  Installing /root/.cargo/bin/cargo-install-update-config
   Installed package `cargo-update v3.0.0` (executables `cargo-install-update`, `cargo-install-update-config`)
```

And the versions of native library packages I installed (beyond `build-essential` and `pkg-config`):
```
root@eoan:/# apt list libss{h,l}{,-dev,*} --installed
Listing... Done
libssl-dev/eoan,now 1.1.1c-1ubuntu4 amd64 [installed]
libssl1.1/eoan,now 1.1.1c-1ubuntu4 amd64 [installed]
```
```
root@bionic:# apt list libss{h,l}{,-dev,*} --installed
Listing... Done
libssl-dev/bionic,now 1.1.0g-2ubuntu4 amd64 [installed]
libssl-doc/bionic,now 1.1.0g-2ubuntu4 all [installed,automatic]
libssl1.1/bionic,now 1.1.0g-2ubuntu4 amd64 [installed]
```

Most notably:
```
root@eoan:/# apt list --installed | grep ssh

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

root@eoan:/#
```
```
root@bionic:/# apt list --installed | grep ssh

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

root@bionic:/#
```

Ubuntu releases:
```
nabijaczleweli@eoan:/$ cat /etc/os-release
NAME="Ubuntu"
VERSION="19.10 (Eoan Ermine)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 19.10"
VERSION_ID="19.10"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=eoan
UBUNTU_CODENAME=eoan
```
```
nabijaczleweli@bionic:~/$ cat /etc/os-release
NAME="Ubuntu"
VERSION="18.04 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic
```

I have read through the issue trackers for our upstreams and didn't see anything related to this, so to further diagnose this I'm asking someone from here with a failing build (if it's anything beyond the usual ubuntu brokenness) to do `apt update && apt full-upgrade`, and post the output of those commands above (`cargo install cargo-update`, `apt list libss{h,l}{,-dev,*} --installed`, `apt list --installed | grep ssh`, `cat /etc/os-release`), to hopefully find a fix for this ASAP.

**Comment by johanwiden** [comment]
Ubuntu 19:10: The command "apt list --installed | grep ssh" listed a number of libraries.
I started to remove them one at a time, and tried installing cargo-update in between (I also discovered that I could no longer install version 2.5.3, which I had managed to install earlier, got the same ld errors), When I finally did "apt remove libssh2-1" I could install cargo-update version 3.0.0.

**Comment by nabijaczleweli** [comment]
Yes! Thanks to your comment I've managed to reproduce this:
```sh
reprepro eoan eoan
mount --bind /proc eoan/proc
chroot eoan
sed -i 's/main/main universe/' /etc/apt/sources.list
apt update
apt install -y --no-install-{suggest,recommend}s pkg-config libssl-dev build-essential wget
bash <(wget -O- https://sh.rustup.rs) -y --profile=minimal
. ~/.cargo/env
cargo install cargo-update     # Works
apt install -y --no-install-{suggest,recommend}s libssh-dev libssh2-1-dev
cargo install -f cargo-update  # Doesn't link
```

Any other combination of `{libssh-dev, libssh2-1-dev}` (`{}`, `{libssh-dev}`, `{libssh2-1-dev}`) links as well. I'll open a bug upstream shortly.

**Comment by nabijaczleweli** [comment]
Apparently this was reported already, but didn't catch my eye.
Tracking upstream bug: https://github.com/alexcrichton/ssh2-rs/issues/173
Related: https://github.com/alexcrichton/ssh2-rs/pull/172, https://github.com/alexcrichton/ssh2-rs/issues/171

Hotfix, mentioned in https://github.com/alexcrichton/ssh2-rs/issues/171, worked for me in the setup above:
Build with `LIBSSH2_SYS_USE_PKG_CONFIG` environment variable set, i.e.
```shell
LIBSSH2_SYS_USE_PKG_CONFIG= cargo install cargo-update
```
for a fresh install.

**Comment by edjacob25** [comment]
Can confirm that using:
```
LIBSSH2_SYS_USE_PKG_CONFIG= cargo install-update -a
```
Works in Ubuntu 18.4.4.

**Comment by VVelox** [comment]
Woot! Just tested here and it worked.

```
env LIBSSH2_SYS_USE_PKG_CONFIG='' cargo install cargo-update
```

**Comment by wez** [comment]
I've published an updated libssh2-sys crate so this should now be resolved

**Comment by nabijaczleweli** [comment]
And an interesting analysis to boot; thanks!

**Comment by nabijaczleweli** [comment]
Released in [`v4.0.0`](https://github.com/nabijaczleweli/cargo-update/releases/tag/v4.0.0).
